### PR TITLE
Type-safe messages - declare io.quarkus.qute.i18n.Message.value() as optional

### LINKED ISSUE #17857: Type-safe messages - consider declaring io.quarkus.qute.i18n.Message.value() as optional
If the relevant properties file is provided, e.g. `msg.properties`.

We would have to validate the presence of the file and key.

### PR DESCRIPTION
**Pull request by michalvavrik**
Resolves: #17857

This PR makes `io.quarkus.qute.i18n.Message#value()`  optional. As a consequence, user can specify message templates for default locale only via localized file (that wasn't possible for Type-safe Message Bundles till now). When user provides message template both via `io.quarkus.qute.i18n.Message#value()` and localized file, `io.quarkus.qute.i18n.Message#value()` value is preferred. Validation is in place, thus if value is not specified at all, exception is thrown. cc @mkouba

### CODE REVIEW

File: extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/MessageBundleProcessor.java [hunk 1]
```diff
@@ -176,29 +179,31 @@ List<MessageBundleBuildItem> processBundles(BeanArchiveIndexBuildItem beanArchiv
 
                     // Find localized files
                     Map<String, Path> localeToFile = new HashMap<>();
+                    // Merge candidates are localized files that are duplicate to localized interface
```
Comment (mkouba) [inline extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/MessageBundleProcessor.java:182]: Hm, WRT "Merge candidates... are duplicate" I think that those localized files are rather _supplementary_ or _additional_.

File: extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/MessageBundleProcessor.java [hunk 2]
```diff
@@ -657,14 +648,81 @@ public String apply(String className) {
...
         }
         return generatedTypes;
     }
 
+    private Map<String, String> getLocalizedFileKeyToTemplate(MessageBundleBuildItem bundle,
+            ClassInfo bundleInterface, String locale, List<MethodInfo> methods, ClassInfo localizedInterface)
+            throws IOException {
+
+        var localizedFile = bundle.getMergeCandidates().get(locale);
+        if (nonNull(localizedFile)) {
```
Comment (mkouba) [inline extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/MessageBundleProcessor.java:663]: What's the point of `nonNull()` here? I mean what's wrong with `localizedFile != null`? ;-)

File: extensions/qute/runtime/src/main/java/io/quarkus/qute/i18n/Message.java [hunk 3]
```diff
@@ -47,9 +53,10 @@
     String key() default DEFAULT_NAME;
 
     /**
+     * This value has higher priority over a message template specified in a localized file.
```
Comment (mkouba) [inline extensions/qute/runtime/src/main/java/io/quarkus/qute/i18n/Message.java:56]: I think that we should mention that 

1. It's a good practice to specify the value,
2. If no value is provided and no localized file matches then the build fails.

**Review by mkouba — COMMENTED** [review]
Looks really good. I've added some minor comments and I also need to take a closer look at the big changeset in the `MessageBundleProcessor`...

File: extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/MessageBundleProcessor.java (hunk 2 above)
Comment (michalvavrik) [reply extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/MessageBundleProcessor.java]: Not really, it reads better to me, but it's irrelevant, I switched to != null and == null

File: extensions/qute/runtime/src/main/java/io/quarkus/qute/i18n/Message.java (hunk 3 above)
Comment (michalvavrik) [reply extensions/qute/runtime/src/main/java/io/quarkus/qute/i18n/Message.java]: sure, thanks

**Comment by michalvavrik** [comment]
@mkouba thank you these are all good points, I pushed changes.

**Review by mkouba — APPROVED** [review]
I think that it would be interesting to know if the build time is not affected significantly by this PR. But we would need some real app sample to test it...

**Comment by mkouba** [comment]
@michalvavrik Merged, thanks!

**Comment by mkouba** [comment]
@michalvavrik Would you care to send a follow-up PR and update the docs accordingly?

**Comment by michalvavrik** [comment]
@mkouba I can add an example and explain `Message#value` is not required and prioritization strategy, if that's what you mean? I wasn't sure previously whether this change is significant enough for user to add this to documentation.

**Pull request merged**